src/main/java/io/fabric8/jenkins/openshiftsync/OpenShiftUtils.java

```diff
@@ -172,445 +175,521 @@ public static boolean isPipelineStrategyBuild(Build b) {
+                for (Route route : routes.getItems()) {
+                    RouteSpec spec = route.getSpec();
+                    if (spec != null
+                            && spec.getTo() != null
+                            && "Service".equalsIgnoreCase(spec.getTo()
+                                    .getKind())
+                            && serviceName.equalsIgnoreCase(spec.getTo()
+                                    .getName())) {
+                        String host = spec.getHost();
+                        if (host != null && host.length() > 0) {
+                            if (spec.getTls() != null) {
+                                return "https://" + host;
+                            }
+                            return "http://" + host;
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Could not find Route for service "
+                        + namespace + "/" + serviceName + ". " + e, e);
+            }
+            // lets try the portalIP instead
+            try {
+                Service service = openShiftClient.services()
+                        .inNamespace(namespace).withName(serviceName).get();
+                if (service != null) {
+                    ServiceSpec spec = service.getSpec();
+                    if (spec != null) {
+                        String host = spec.getClusterIP();
+                        if (host != null && host.length() > 0) {
+                            return defaultProtocolText + host;
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Could not find Route for service "
+                        + namespace + "/" + serviceName + ". " + e, e);
+            }
+        }
+
+        // lets default to the service DNS name
+        return defaultProtocolText + serviceName;
+    }
+
+    /**
+     * Calculates the external URL to access Jenkins
+     *
+     * @param namespace
+     *            the namespace Jenkins is runing inside
+     * @param openShiftClient
+     *            the OpenShift client
+     * @return the external URL to access Jenkins
+     */
+    public static String getJenkinsURL(OpenShiftClient openShiftClient,
+            String namespace) {
+        // if the user has explicitly configured the jenkins root URL, use it
+        String rootUrl = Jenkins.getInstance().getRootUrl();
+        if (StringUtils.isNotEmpty(rootUrl)) {
+            return rootUrl;
+        }
+
+        // otherwise, we'll see if we are running in a pod and can infer it from
+        // the service/route
+        // TODO we will eventually make the service name configurable, with the
+        // default of "jenkins"
+        return getExternalServiceUrl(openShiftClient, "http://", namespace,
+                "jenkins");
+    }
+
+    public static String getNamespacefromPodInputs() {
```

Comment:
(in case my previous comment on the outdated commit gets lost):

i think this should cache the value (or should only be called once, during plugin initialization, rather than per build update as is currently the case). The pod's namespace can't change.

Reply:
ahh ... should have formatted before making the change ;-)

Reply:
cache update pushed

Reply:
cache update 2 pushed (squashed with original cache update)

Reply:
using static init